### DISCUSSION TITLE
How to resolve symlink to full path in Set-Location (cd)

### DISCUSSION THREAD

**Discussion by AFKBETO** [PowerShell/PowerShell#21159] [Q&A]
I would like to know how to physically switch to the full path of the target location of a symlink.

Example, I have a symlink ~\folder => C:\\parent\folder

When I am at ~, ``cd <option> .\folder`` should physically put me in ``C:\\parent\folder`` instead of ``~\folder.`` This means that:
- ``pwd`` should return ``C:\\parent\folder`` and not ``$HOME\folder``
- ``cd ..`` should bring me to ``C:\\parent`` and not back to ~

I could not find any mention in the documentation, so idk if there is anyway to switch location like that without using the full path.

Many thanks!

**Comment by rhubarb-geek-nz** [comment]
I have had a play with this on Linux because I can't use mklink to create a directory symbolic link on Windows with out admin rights.

Firstly I don't know what you mean by tilde, do you mean the actual character or the value of $HOME ? Are you using it as a place holder?

There are two issues at play here...

1. PowerShell uses a logical Location, not the real current directory. So it simulates changing directory by holding a value that represents the current directory.

2. You can go from filename to file, or from filename to directory, but you can't guarantee going from file or directory back to filename because multiple filenames can point to the same file or directory.

If you want to work in a different directory and then return back to the same directory I recommend using Push-Location and Pop-Location

If you are in a directory and Get-Location does not give you a valid path to the directory then how did you get there?

**Reply by AFKBETO** [reply]
Thanks for the answer!

Tilde is basically $HOME. Sorry for using both in the OP.

The behavior of Set-Location I stated in the OP means that all commands I use after Set-Location would be in ``$HOME\folder`` and not in ``C:\\parent\folder``, where the symlink points to. For most use cases, these yield identical results. However, some applications can have unexpected behavior (like https://github.com/nodejs/node/issues/34866). As such, I kind of need some methods to navigate in powershell that resolve to the target location, and then launch these problematic applications there.

While I can use the absolute path instead, it is, you know, not ideal.

**Comment by rhubarb-geek-nz** [comment]
Ignoring PowerShell, it is consistent with command line

```
C:\Windows\System32>mkdir \parent

C:\Windows\System32>mkdir \parent\folder

C:\Windows\System32>cd \Users\bythesea

C:\Users\bythesea>mklink /D folder c:\parent\folder
symbolic link created for folder <<===>> c:\parent\folder

C:\Users\bythesea>cd folder

C:\Users\bythesea\folder>cd ..

C:\Users\bythesea>
```

Although it is a symbolic link, the current directory did become $HOME\folder

And with PowerShell

```
C:\Users\bythesea>pwsh
PowerShell 7.4.1
PS C:\Users\bythesea> Get-Item folder

    Directory: C:\Users\bythesea

Mode                 LastWriteTime         Length Name
----                 -------------         ------ ----
l----          30/01/2024    16:37                folder -> c:\parent\folder

PS C:\Users\bythesea> Set-Location folder
PS C:\Users\bythesea\folder> Get-Location

Path
----
C:\Users\bythesea\folder

PS C:\Users\bythesea\folder> Set-Location $HOME
PS C:\Users\bythesea> push-location folder
PS C:\Users\bythesea\folder> get-location

Path
----
C:\Users\bythesea\folder

PS C:\Users\bythesea\folder> pop-location
PS C:\Users\bythesea>
```

And consistent with bash on Linux

```
$ sudo mkdir /parent
$ sudo mkdir /parent/folder
$ ln -s /parent/folder folder
$ ls -ld folder
lrwxrwxrwx 1 bythesea users 14 Jan 30 17:04 folder -> /parent/folder
$ cd folder
$ pwd
/home/bythesea/folder
```

**Reply by jgargus** [reply]
They didn't ask "is this consistent with bash". They asked for a solution on how to change to the real directory, which is possible with bash on Linux.

Notably, pwd (the bash shell built-in) and /bin/pwd (actual program) give different answers as to the location. So, you can use "cd \`/bin/pwd\`" to change to the real directory at any time.

`$ ls -l src
lrwxrwxrwx 1 me ... src -> /c/src/

$ cd src

$ pwd
/c/Users/me/src

$ /bin/pwd
/c/src

$ cd \`/bin/pwd\`

$ pwd
/c/src`

They (and myself) are asking for a way, if possible, to do this in Powershell, because some things break when you are under a symlink.

**Comment by rhubarb-geek-nz** [comment]
In answer to the original question, I hope....

```
PS> (Get-Item folder).Target
c:\parent\folder
```

**Reply by jgargus** [reply]
While this technically does let someone see the target of the symbolic link, it doesn't really solve the problem.

What OP (and I) want is to be able to either have cd detect and do this automatically OR for there to be a way to change to the real directory after navigating through the symlink.

On Linux/Unix and with cygwin on Windows, I used to do this all the time using an alias I had set to "cd \`/bin/pwd\`", because /bin/pwd (unlike the bash shell built-in pwd) would always print out the real directory path.

So, I think people are searching for some command or script to do this, so we get the benefits of using symlinks to make things easy to find/navigate, but don't run into things breaking because we followed a symlink instead of a real folder path.
